module3/homework/advanced_calculator/advancedCalculator.hpp

```diff
@@ -0,0 +1,52 @@
+
+#pragma once
+#include <cmath>
+#include <functional>
+#include <map>
+#include <regex>
+#include <string>
+
+namespace Pattern {
+const auto expression = R"(^([-]?\d+\.\d+|[-]?\d+)\s?([+*/\-%^$])\s?([-]?\d+\.\d+|[-]?\d+))";
+const auto factiorial = R"(([-]?\d+\.\d+|[-]?\d+)\s?(\!))";
+}  // namespace Pattern
+
+enum OperationType {
+    TwoArguments,
+    Factorial,
+    Bad
+};
+
+enum ErrorCode {
+    OK,
+    BadCharacter,
+    BadFormat,
+    DivideBy0,
+    SqrtOfNegativeNumber,
+    ModuleOfNonIntegerValue
+};
+
+const std::map<char, std::function<double(double, double)>> operations{
+    {'+', std::plus<double>()},
+    {'-', std::minus<double>()},
+    {'*', std::multiplies<double>()},
+    {'/', std::divides<double>()},
+    {'%', std::modulus<int>()},
+    {'^', [](double base, double pow) { return std::pow(base, pow); }},
+    {'$', [](double base, double root) { return std::pow(base, 1.0 / root); }}};
```

Comment:
You could put lambda for factorial there.

Reply:
+1